### ISSUE TITLE
Cannot Call Object Methods From Coroutines

### ISSUE THREAD

**Issue by PatSmuk** [ElunaLuaEngine/Eluna#126]
**TL;DR:** Due to the way method calls happen in Eluna, the wrong Lua context is being used by the method implementations. I want to pass the correct one through every method call.
## Proof of Concept

Test code (save in `test.lua`, run with `.test`):

``` lua
local function command_handler(event, player, command)
    if command ~= 'test' then
        return true
    end

    player:SendUnitSay("Test outside!", 0)

    local c = coroutine.create(function ()
        player:SendUnitSay("Test inside!", 0)
    end)
    coroutine.resume(c)

    return false
end

RegisterPlayerEvent(42, command_handler)
```

**Expected behavior:** player says "Test outside!" followed by "Test inside!".
**Actual behavior:** player says "Test outside!" and the following Lua error is printed to the console:

```
lua_scripts/test.lua:11: bad argument #2 to 'resume' (string expected, got no value)
```
## Explanation of Bug

`coroutine.resume` does not take a string as it's second argument; what's happening is that SendUnitSay is called but tries to pull arguments off the main stack instead of the coroutine's stack. 

The last thing on the main stack is `coroutine.resume(c)`, which had no second or third parameter, so SendUnitSay fails. Since the error reporter also uses the wrong stack, it incorrectly reports `coroutine.resume` as the source of the error.

Eluna was designed under the assumption that only one `lua_State` exists. This is incorrect: whenever you create a new coroutine Lua creates a new corresponding `lua_State`.
## Proposed Solution

`ElunaTemplate::thunk` and `ElunaGlobal::thunk` receive the correct `lua_State`, but they don't pass it to their target. My proposed solution is to simply pass it to every Eluna function alongside the `Eluna` object:

``` cpp
int expected = l->mfunc(E, obj);
```

becomes:

``` cpp
int expected = l->mfunc(E, L, obj);
```

As for the methods:

``` cpp
int GetCaster(Eluna* E, Aura* aura)
{
    Eluna::Push(E->L, aura->GetCaster());  // E->L is *WRONG* when this is called from a coroutine!!!
    return 1;
}
```

becomes:

``` cpp
int GetCaster(Eluna* /*E*/, lua_State *L, Aura* aura)
{
    Eluna::Push(L, aura->GetCaster());
    return 1;
}
```
## Alternative Solutions

Have `Eluna::L` be updated dynamically to point to the current context. This would require more work and potentially be error-prone.
